pkg/app/instances/docker.go

```diff
@@ -297,3 +310,31 @@ func (m *DockerInstanceManager) getContainerLabel(host string, key string) (stri
 	}
 	return value, nil
 }
+
+func (m *DockerInstanceManager) downloadDockerImageIfNeeded(ctx context.Context) error {
+	listRes, err := m.Client.ImageList(ctx, image.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("Failed to list docker images: %w", err)
+	}
+	for _, image := range listRes {
+		for _, tag := range image.RepoTags {
+			if tag == m.Config.Docker.DockerImageName {
+				return nil
+			}
+		}
+	}
+
+	reader, err := m.Client.ImagePull(ctx, m.Config.Docker.DockerImageName, image.PullOptions{})
+	if err != nil {
+		return fmt.Errorf("Failed to request pulling docker image '%s': %w", m.Config.Docker.DockerImageName, err)
+	}
+	defer reader.Close()
+	// Caller of ImagePull should handle its output to complete actual ImagePull operation.
+	// Details in https://pkg.go.dev/github.com/docker/docker/client#Client.ImagePull.
+	_, err = io.Copy(io.Discard, reader)
+	if err != nil {
+		return fmt.Errorf("Failed to pull docker image '%s': %w", m.Config.Docker.DockerImageName, err)
```

Comment:
nit: IIRC if using format verb `%q` rather than `%s` you don't have to add the quotes yourself.

Reply:
Done, thanks for the review!